pkg/activator/throttler.go

```diff
@@ -0,0 +1,184 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package activator
+
+import (
+	"github.com/knative/serving/pkg/queue"
+	"go.uber.org/zap"
+	"sync"
+	"github.com/knative/serving/pkg/client/informers/externalversions/serving/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+	v1alpha12 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"errors"
+	corev1 "k8s.io/api/core/v1"
+	. "github.com/knative/serving/pkg/reconciler"
+	"github.com/knative/serving/pkg/reconciler"
+	"fmt"
+)
+
+// NewThrottler creates a new Throttler
+func NewThrottler(queueDepth, maxConcurrency, initialCapacity int32, revisionInformer v1alpha1.RevisionInformer, logger *zap.SugaredLogger) *Throttler {
+	breakers := make(map[RevisionID]queue.Breaker)
+	services := make(map[RevisionID]string)
+	return &Throttler{breakers: breakers, services: services, queueDepth: queueDepth, maxConcurrency: maxConcurrency, initialCapacity: initialCapacity, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler allows to update the max concurrency of registered Breakers
+type Throttler struct {
+	breakers         map[RevisionID]queue.Breaker
+	services         map[RevisionID]string
+	revisionInformer v1alpha1.RevisionInformer
+	queueDepth       int32
+	maxConcurrency   int32
+	initialCapacity  int32
+	logger           *zap.SugaredLogger
+	mux              sync.RWMutex
+}
+
+// Add potentially creates a new breaker for the revision
+// returns true if a new entry was added
+func (t *Throttler) AddIfAbsent(rev RevisionID) bool {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	if _, ok := t.breakers[rev]; !ok {
+		breaker := queue.NewBreaker(t.queueDepth, t.maxConcurrency, t.initialCapacity)
+		t.breakers[rev] = *breaker
+		t.services[rev] = GetServingK8SRevNameForObj(rev.Name)
+		return true
+	}
+	return false
+}
+
+// Exists returns true if a Breaker already exists for the given revision
+func (t *Throttler) Exists(rev RevisionID) bool {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	_, ok := t.breakers[rev]
+	return ok
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity update the max concurrency of the Breaker corresponding to the revision
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.RLock()
+	defer t.mux.RUnlock()
+	var containerConcurrency int32
+	serviceName, ok := t.services[rev]
+	if !ok {
+		return fmt.Errorf("no service name registered for the revision: %s", rev)
+	}
+	revision, err := t.getRevision(rev.Namespace, serviceName, t.revisionInformer)
+	if err != nil {
+		return fmt.Errorf("no revision was found for: %s", rev)
+	}
+	containerConcurrency = int32(revision.Spec.ContainerConcurrency)
+	// Equals to the pending queue size as defined in `cmd/queue/main.go`
+	if containerConcurrency < 10 {
+		containerConcurrency = 10
+	}
```

Comment:
We should decide how much we want to let through here.

1. If we only want to let through as much as the revision can currently handle, we should just let `size * containerConcurrency` through, without bounding it to 10.
2. If we want to let through as much as the revision can buffer, we should calculate: `size (pendinQueue + containerConcurrency)`.

The current implementation seems a bit in between.